packages/unpublished-eslint-rules/index.js

```diff
@@ -0,0 +1,285 @@
+                      changes.push(fixer.insertTextBeforeRange(declaration.range, newLines));
+                    }
+                    removals.forEach(range => {
+                      changes.push(fixer.removeRange(range));
+                    });
+                    return changes;
+                  },
+                });
+              }
+            }
+            // console.log('\nProgram:exitEnd\n======\n\n\n');
+          },
+          ImportDeclaration(node) {
+            declarations.push(node);
+            lastImport = node;
+          },
+          ImportDefaultSpecifier(node) {
+            imports[node.local.name] = {
+              local: node.local,
+              imported: node.imported || node.local,
+              from: node.parent.source,
+              isDefault: true,
+            };
+          },
+          ImportSpecifier(node) {
+            imports[node.local.name] = {
+              local: node.local,
+              imported: node.imported || node.local,
+              from: node.parent.source,
+              isDefault: false,
+            };
+          },
+          Identifier(node) {
+            let type = node.parent.type;
+
+            if (type === 'ImportSpecifier' || type === 'ImportDefaultSpecifier') {
+              // console.log(`found ${node.name} for import type ${type}`);
+            } else if (type === 'TSTypeReference') {
+              // console.log(`found type reference ${node.name}`);
+              // do nothing, this Identifier is handled elsewhere
+            } else if (imports[node.name] !== undefined) {
+              const scope = context.getScope();
+              if (scope.type === 'class') {
+                usedImports[node.name] = true;
+              } else {
+                // determine if this Identifier is ours declared via an import
+                // (vs. potentially a shadowed variable instance)
+                for (let i = 0; i < scope.references.length; i++) {
+                  let ref = scope.references[i];
+                  if (ref.identifier === node) {
+                    if (ref.resolved.scope.type === 'module') {
+                      // console.log(`found import usage ${node.name} in ${node.parent.type}`);
+                      usedImports[node.name] = true;
+                      break;
+                    }
+                  }
+                }
+              }
+            }
+          },
+          TSTypeReference(node) {
+            let identifier = node.typeName;
+            if (imports[identifier.name] !== undefined) {
+              // console.log(`found used type ${identifier.name}`);
+              if (node.typeParameters && node.typeParameters.params.length) {
+                // this type usage contains generics e.g. `Dict<Thing>`
+                // so we save off how many params it uses to fix up the import
+                // later
+                usedTypes[identifier.name] = node.typeParameters.params.length;
```

Comment:
Maybe create another object to keep the type params so we don't have both booleans and numbers in this one